File: Algorithm.CSharp/FutureOptionModelsConsistencyRegressionAlgorithm.cs
/*
 * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
 * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

using System;
using System.Collections.Generic;
using System.Linq;
using QuantConnect.Data;
using QuantConnect.Securities;

namespace QuantConnect.Algorithm.CSharp
{
    /// <summary>
    /// Regression algorithm asserting that models set to a future option are passed down to the chain contracts
    /// </summary>
    public class FutureOptionModelsConsistencyRegressionAlgorithm : OptionModelsConsistencyRegressionAlgorithm
    {
        public override void Initialize()
        {
            SetStartDate(2012, 1, 3);
            SetEndDate(2012, 1, 4);

            var dcContract = AddFutureContract(
                QuantConnect.Symbol.CreateFuture(
                    Futures.Dairy.ClassIIIMilk,
                    Market.CME,
                    new DateTime(2012, 4, 1)),
                Resolution.Daily);

            var optionSymbol = OptionChainProvider.GetOptionContractList(dcContract.Symbol, Time).First();

            AddFutureOption(optionSymbol.Canonical);
            SetModels(Securities[optionSymbol.Canonical]);

            var optionContract = AddFutureOptionContract(optionSymbol, Resolution.Daily);

            CheckModels(optionContract);
        }
        public override void OnData(Slice slice)
        {
            // Empty, we just don't want the base class checks done in OnData since we are adding the option contract in Initialize
            // and checking the models.
        }

        /// <summary>
        /// This is used by the regression test system to indicate if the open source Lean repository has the required data to run this algorithm.
        /// </summary>
        public override bool CanRunLocally => true;

        /// <summary>
        /// This is used by the regression test system to indicate which languages this algorithm is written in.
        /// </summary>
        public override Language[] Languages { get; } = { Language.CSharp, Language.Python };

        /// <summary>
        /// Data Points count of all timeslices of algorithm
        /// </summary>
        public override long DataPoints => 76883;

        /// <summary>
        /// Data Points count of the algorithm history
        /// </summary>
        public override int AlgorithmHistoryDataPoints => 0;

        /// <summary>
        /// This is used by the regression test system to indicate what the expected statistics are from running the algorithm
        /// </summary>
        public override Dictionary<string, string> ExpectedStatistics => new Dictionary<string, string>
        {
            {"Total Trades", "0"},
            {"Average Win", "0%"},
            {"Average Loss", "0%"},
            {"Compounding Annual Return", "0%"},
            {"Drawdown", "0%"},
            {"Expectancy", "0"},
            {"Net Profit", "0%"},
            {"Sharpe Ratio", "0"},
            {"Probabilistic Sharpe Ratio", "0%"},
            {"Loss Rate", "0%"},
            {"Win Rate", "0%"},
            {"Profit-Loss Ratio", "0"},
            {"Alpha", "0"},
            {"Beta", "0"},
            {"Annual Standard Deviation", "0"},
            {"Annual Variance", "0"},
            {"Information Ratio", "0"},
            {"Tracking Error", "0"},
            {"Treynor Ratio", "0"},
            {"Total Fees", "$0.00"},
            {"Estimated Strategy Capacity", "$0"},
            {"Lowest Capacity Asset", ""},
            {"Portfolio Turnover", "0%"},
            {"OrderListHash", "d41d8cd98f00b204e9800998ecf8427e"}
        };
    }
}

File: Algorithm.Python/FutureOptionModelsConsistencyRegressionAlgorithm.py
# QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
# Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from AlgorithmImports import *

from OptionModelsConsistencyRegressionAlgorithm import OptionModelsConsistencyRegressionAlgorithm

### <summary>
### Regression algorithm asserting that models set to a future option are passed down to the chain contracts
### </summary>
class FutureOptionModelsConsistencyRegressionAlgorithm(OptionModelsConsistencyRegressionAlgorithm):

    def Initialize(self) -> None:
        self.SetStartDate(2012, 1, 3)
        self.SetEndDate(2012, 1, 4)

        dc_contract = self.AddFutureContract(
            Symbol.CreateFuture(
                Futures.Dairy.ClassIIIMilk,
                Market.CME,
                datetime(2012, 4, 1)),
            Resolution.Daily)

        option_symbol = list(self.OptionChainProvider.GetOptionContractList(dc_contract.Symbol, self.Time))[0]

        self.AddFutureOption(option_symbol.Canonical)
        self.SetModels(self.Securities[option_symbol.Canonical])

        option_contract = self.AddFutureOptionContract(option_symbol, Resolution.Daily)

        self.CheckModels(option_contract)

    def OnData(self, slice: Slice) -> None:
        # Empty, we just don't want the base class checks done in OnData since we are adding the option contract in Initialize
        # and checking the models.
        pass
